File: day3.py
import common


def num_bits_set(all_numbers: list[str]) -> list[int]:
    bit_counts = [0] * len(all_numbers[0])
    for number in all_numbers:
        for i in range(len(number)):
            if number[i] == '1':
                bit_counts[i] = bit_counts[i] + 1
    return bit_counts


def num_bits_set_at_position(all_numbers: list[str], position: int) -> int:
    bit_count = 0
    for number in all_numbers:
        if number[position] == '1':
            bit_count = bit_count + 1
    return bit_count


def power_consumption(all_numbers: list[str]) -> int:
    all_numbers_len = len(all_numbers)
    counts = num_bits_set(all_numbers)
    gamma_rate = epsilon_rate = 0
    for count in counts:
        gamma_rate = gamma_rate << 1
        epsilon_rate = epsilon_rate << 1
        if count > all_numbers_len / 2:
            gamma_rate = gamma_rate + 1
        else:
            epsilon_rate = epsilon_rate + 1

    return gamma_rate * epsilon_rate


def life_support_rating(all_numbers: list[str]) -> int:
    oxygen_generator_numbers = all_numbers
    co2_scrubber_numbers = all_numbers

    for bit_position in range(len(all_numbers[0])):
        print(bit_position, len(oxygen_generator_numbers), len(co2_scrubber_numbers))
        if len(oxygen_generator_numbers) == 0 and len(co2_scrubber_numbers) == 0:
            break

        if len(oxygen_generator_numbers) > 1:
            # aka '1' is more common or same
            if num_bits_set_at_position(oxygen_generator_numbers, bit_position) >= len(oxygen_generator_numbers) / 2:
                oxygen_generator_numbers = [n for n in oxygen_generator_numbers if n[bit_position] == '1']
            else:
                oxygen_generator_numbers = [n for n in oxygen_generator_numbers if n[bit_position] == '0']

        if len(co2_scrubber_numbers) > 1:
            # aka '1' is more common or same
            if num_bits_set_at_position(co2_scrubber_numbers, bit_position) >= len(co2_scrubber_numbers) / 2:
                co2_scrubber_numbers = [n for n in co2_scrubber_numbers if n[bit_position] == '0']
            else:
                co2_scrubber_numbers = [n for n in co2_scrubber_numbers if n[bit_position] == '1']

    return int(oxygen_generator_numbers[0], 2) * int(co2_scrubber_numbers[0], 2)


def main():
    print("day 3 part 1:", power_consumption(common.lines_in_file("day3.dat")))
    print("day 3 part 2:", life_support_rating(common.lines_in_file("day3.dat")))


if __name__ == '__main__':
    main()

File: src/main/java/adventofcode2021/Day3.java
package adventofcode2021;

import static adventofcode2021.Common.inputAsArrayFor;

import java.util.Arrays;
import java.util.stream.Collectors;

public class Day3 {

	public static long lifeSuport(String[] numbers) {
		String[] oxygenGeneratorNumbers = numbers;
		String[] co2ScrubberNumbers = numbers;

		for (int bitPosition = 0; bitPosition < numbers[0].length(); ++bitPosition) {
			final int safeBitPosition = bitPosition;
			if (oxygenGeneratorNumbers.length > 1) {
				// note that if the array length is odd, then len/2 is less than half. "most"
				// means one more than len/2 in that case
				if (numBitsSetAtPostion(oxygenGeneratorNumbers,
						safeBitPosition) >= (oxygenGeneratorNumbers.length / 2 + oxygenGeneratorNumbers.length % 2)) {
					oxygenGeneratorNumbers = Arrays.stream(oxygenGeneratorNumbers)
							.filter(n -> n.charAt(safeBitPosition) == '1').toArray(String[]::new);
				} else {
					oxygenGeneratorNumbers = Arrays.stream(oxygenGeneratorNumbers)
							.filter(n -> n.charAt(safeBitPosition) == '0').toArray(String[]::new);
				}
			}
			if (co2ScrubberNumbers.length > 1) {
				// note that if the array length is odd, then len/2 is less than half. "most"
				// means one more than len/2 in that case
				if (numBitsSetAtPostion(co2ScrubberNumbers,
						safeBitPosition) >= (co2ScrubberNumbers.length / 2 + co2ScrubberNumbers.length % 2)) {
					co2ScrubberNumbers = Arrays.stream(co2ScrubberNumbers).filter(n -> n.charAt(safeBitPosition) == '0')
							.toArray(String[]::new);
				} else {
					co2ScrubberNumbers = Arrays.stream(co2ScrubberNumbers).filter(n -> n.charAt(safeBitPosition) == '1')
							.toArray(String[]::new);
				}
			}
		}

		return Long.valueOf(oxygenGeneratorNumbers[0], 2) * Long.valueOf(co2ScrubberNumbers[0], 2);
	}

	static public int[] numBitsSet(String[] allNumbers) {
		int[] bitCounts = new int[allNumbers[0].length()];
		for (String number : allNumbers) {
			for (int i = 0; i < number.length(); i++) {
				if (number.charAt(i) == '1')
					bitCounts[i] += 1;
			}
		}
		return bitCounts;
	}

	public static int numBitsSetAtPostion(String[] numbers, int bitPosition) {
		int bitCount = 0;
		for (String number : numbers)
			if (number.charAt(bitPosition) == '1')
				bitCount += 1;
		return bitCount;
	}

	public static long powerConsumption(String[] lines) {
		int[] numBitsSet = numBitsSet(lines);
		String gammaRateString = Arrays.stream(numBitsSet).mapToObj(n -> n >= lines.length / 2)
				.map(mostCommon -> mostCommon ? "1" : "0").collect(Collectors.joining());
		String epsilonRateString = "";
		for (char c : gammaRateString.toCharArray())
			epsilonRateString += c == '1' ? "0" : "1";
		long powerConsumption = Long.valueOf(gammaRateString, 2) * Long.valueOf(epsilonRateString, 2);
		return powerConsumption;
	}

	public static void main(String[] args) {
		System.out.println("Day 3 part 1: " + powerConsumption(inputAsArrayFor("day3")));
		System.out.println("Day 3 part 2: " + lifeSuport(inputAsArrayFor("day3")));
	}
}
